### PR TITLE
fix: chain parameters should inspect conditions to narrow down possible target types

### PR DESCRIPTION
**Pull request by ssvegaraju**
Issue #166, if available:

Description of changes:
Added logic for parsing compiledSearchParamaters' conditions to narrow down target Resource types when there are multiple, if possible.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/QueryBuilder/chain.ts [hunk 1]
```diff
@@ -55,9 +55,28 @@ const parseChainedParameters = (
                         );
                     }
                 } else if (fhirSearchParam.target?.length !== 1) {
-                    throw new InvalidSearchParameterError(
-                        `Chained search parameter '${searchModifier.parameterName}' for resource type ${currentResourceType} points to multiple resource types, please specify.`,
-                    );
+                    // check compiled[].condition for resolution
+                    if (fhirSearchParam.compiled.length > 0) {
+                        const compiled = fhirSearchParam.compiled.pop()!; // we can use ! since we checked length before
```
Comment (carvantes) [inline src/QueryBuilder/chain.ts]: I don't think it's safe to use `pop` here since it mutates the array. `fhirSearchParam` comes from `fhirSearchParametersRegistry` and calling `pop` will permanently remove `compiled` items from the array and cause trouble in future requests.

File: src/QueryBuilder/chain.test.ts [hunk 2]
```diff
@@ -62,6 +62,30 @@ describe('parseChainedParameters', () => {
               },
             ]
         `);
+
+        expect(
+            parseChainedParameters(fhirSearchParametersRegistry, 'DocumentReference', {
```
Comment (carvantes) [inline src/QueryBuilder/chain.test.ts]: move this to a separate test. Name it something like "search param with conditions that narrow down target type"

File: src/QueryBuilder/chain.ts [hunk 3]
```diff
@@ -55,9 +55,28 @@ const parseChainedParameters = (
...
+                        // if there is no resolve condition, we have multiple resources pointed to.
+                        // condition's format is defined in `../FHIRSearchParamtersRegistry/index.ts`
+                        if (!compiled.condition || compiled.condition[1] !== COMPILED_CONDITION_OPERATOR_RESOLVE) {
+                            throw new InvalidSearchParameterError(
+                                `Chained search parameter '${searchModifier.parameterName}' for resource type ${currentResourceType} points to multiple resource types, please specify.`,
```
Comment (carvantes) [inline src/QueryBuilder/chain.ts]: Refactor so that this error is thrown only once. It makes updating the error message later easier.

In general it'd be better to move most of the new code in this PR to a separate function. Maybe something like `getUniqueTarget(fhirSearchParam): string`

File: src/QueryBuilder/chain.ts (hunk 1 above)
Comment (carvantes) [reply src/QueryBuilder/chain.ts]: What happens if there are multiple `compiled` ? 
I think that we should only narrow down the type after verifying that ALL `compiled` have a condition that resolves to a type.
Sorry if my earlier slack message wasn't clear.

**Review by carvantes — CHANGES_REQUESTED** [review]

File: src/QueryBuilder/chain.test.ts (hunk 2 above)
Comment (ssvegaraju) [reply src/QueryBuilder/chain.test.ts]: I also added tests for the edge cases we went over!

File: src/QueryBuilder/chain.ts (hunk 1 above)
Comment (ssvegaraju) [reply src/QueryBuilder/chain.ts]: Thanks for helping clear up the confusion on the edge cases, I refactored to a method and removed the mutation of `fhirSearchParametersRegistry`

File: src/QueryBuilder/chain.ts [hunk 4]
```diff
@@ -1,14 +1,36 @@
...
+import { FHIRSearchParametersRegistry, SearchParam } from '../FHIRSearchParametersRegistry';
+import { COMPILED_CONDITION_OPERATOR_RESOLVE, NON_SEARCHABLE_PARAMETERS } from '../constants';
 import { parseSearchModifiers, normalizeQueryParams, isChainedParameter } from '../FhirQueryParser/util';
 
 export interface ChainParameter {
     chain: { resourceType: string; searchParam: string }[];
     initialValue: string[];
 }
 
+export function getUniqueTarget(fhirSearchParam: SearchParam): string | undefined {
```
Comment (carvantes) [inline src/QueryBuilder/chain.ts:12]: Can you move the rest of the existing logic regarding finding target into this function?
I think that this function should stand on its own and give us the unique target no matter what. Right now the function assumes that the `fhirSearchParam.target?.length !== 1` check was done outside.

File: src/QueryBuilder/chain.test.ts [hunk 5]
```diff
@@ -93,4 +93,98 @@ describe('parseChainedParameters', () => {
...
+                {
+                    resourceType: 'Person',
+                    path: 'link.target',
+                    condition: ['link.something', 'resolve', 'Patient'],
+                },
+            ],
+        };
+
+        expect(getUniqueTarget(successCase)).toMatchInlineSnapshot(`"Patient"`);
+        errorCases.forEach((errorCase) => expect(getUniqueTarget(errorCase)).toMatchInlineSnapshot(`undefined`));
```
Comment (carvantes) [inline src/QueryBuilder/chain.test.ts]: Use `.toBeUndefined()` instead. 

`toMatchInlineSnapshot` should not be used inside loops(since it needs to write potentially different snapshots for each tests). I think that In this particular case it works since they all have the same snapshot and you wrote the snapshot manually. AFAIK Jest crashes if you attempt to update that snapshot in a loop

File: src/QueryBuilder/chain.ts (hunk 4 above)
Comment (carvantes) [reply src/QueryBuilder/chain.ts:12]: consider the case where the conditions resolve to a type that is NOT in the target array(It's unlikely since that means the search parameter definition is messed up). It should return undefined in that case. e.g:
```json
{
    "name": "patient",
    "url": "http://hl7.org/fhir/SearchParameter/aaa",
    "type": "reference",
    "description": "...",
    "base": "Person",
    "target": [
      "Patient",
      "Practitioner"
    ],
    "compiled": [
      {
        "resourceType": "Person",
        "path": "link.target",
        "condition": [
          "link.target",
          "resolve",
          "Observation"
        ]
      },
      {
        "resourceType": "Person",
        "path": "link.target",
        "condition": [
          "link.something",
          "resolve",
          "Observation"
        ]
      }
    ]
  },
```

**Review by carvantes — CHANGES_REQUESTED** [review]

File: src/QueryBuilder/chain.ts (hunk 4 above)
Comment (ssvegaraju) [reply src/QueryBuilder/chain.ts:12]: Added this new edge case, and pulled out the if statement in parseChainedParameters.

File: src/QueryBuilder/chain.test.ts (hunk 5 above)
Comment (ssvegaraju) [reply src/QueryBuilder/chain.test.ts]: I did not know that about jest, thanks for letting me know! Amended the test cases to use toBeUndefined.

**Review by carvantes — APPROVED** [review]
